web/packages/teleterm/src/ui/Search/SearchContext.tsx

```diff
@@ -50,6 +51,7 @@ export const SearchContextProvider: FC = props => {
   const previouslyActive = useRef<Element>();
   const inputRef = useRef<HTMLInputElement>();
   const [isOpen, setIsOpen] = useState(false);
+  const [isLockedOpen, setIsLockedOpen] = useState(false);
```

Comment:
I'm wondering if this could be `useRef`. Theoretically UI doesn't depend on this value, so we could avoid additional re-renders when `setIsLockedOpen` called.

Reply:
That's a good idea. I thought it's weird that all those `close` functions will change when `isLockedOpen` gets updated but I didn't think of using a ref here.

Reply:
Also it made me realize that `lockOpen` is not safe for concurrent use. I'll just leave a comment about it since it's meant to be used in situations where the user is locked out of interacting with the search bar anyway (such as when opening a modal).

Reply:
Locking doesn't work properly with `useRef` 😬 - when you click 'Close' on the modal, the search closes.
I see two options here: add `capture` flag to the event listener that listens for an outside click, so the `SearchContext.close` is called before `onCancel` on the modal , or wrap `isLockedOpenRef.current = false;` in `setTimeout` (but I'm not sure if this will be free from race conditions).

Reply:
Using `capture` seems like a better approach out of those two, but even with `capture` pressing Escape will close both the modal and the search bar. Whereas the previous solution with `useState` would work correctly in this case.

So I guess perhaps using `useState` is warranted in this scenario? I'm just worried that with `useState`, we're dependent on how state interacts with event handlers.

I suppose we could add a test with an actual ModalsHost and SearchBar to verify that the behavior stays correct?

---

Even then, it's unlikely that the React team will change this without us knowing:

> React waits until all code in the event handlers has run before processing your state updates.

https://react.dev/learn/queueing-a-series-of-state-updates

So using state in this case kinda makes sense. Basically once `lockOpen` is called we update the state, so that all event handlers process with that state in mind. Once they're processed, the state is updated and the behavior of `close` switches.

For me, that's easier to think about than using a ref and worrying about the order of native events.

Reply:
Ok, I didn't know that React waits for all event handlers, it makes sense to bring back `useState`.

Reply:
In the end I will need to add `capture` to the event listener anyway. When `lockOpen` is called outside of an event listener, such as right now when I'm implementing `retryWithRelogin`-like behavior, `lockOpen` removes the lock first and then the events are fired so it just doesn't work.

The issue I mentioned with Escape closing both the modal and the search bar is not present with `capture` when `useState` is used.